design/Robust/empirical.Rmd

```diff
@@ -0,0 +1,124 @@
+
+```{r}
+cat(ret$LST)
+```
+
+# Implementations in `mmrm`
+
+Since using the "empirical" option, we are changing all standard errors and test statistics involving the
+fixed effect parameters, `beta_vcov`. So it can be similar to our other methods, "Satterthwaite" and "Kenward-Roger".
+So in `mmrm`, the method, stands for both "covariance adjust" and "degree of freedom calculation".
+In SAS, they are separated arguments.
+
+covariance:
+
+1. traditional ${X^\top \Omega^{-1} X}^{-1}$
+1. empirical ${X^\top \Omega^{-1} X}^{-1} \sum_{s}{X_i^\top \Omega^{-1} \epsilon_i \epsilon_i^\top \Omega^{-1} X_i} {X^\top \Omega^{-1} X}^{-1}$
+1. Kenward-Roger(see KR documentation)
+
+degree of freedom
+
+1. contain (not implemented)
+1. between within (default one)
+1. residual (not implemented)
+1. Kenward-Roger (this will always use the KR covariance matrix)
+1. Satterthwaite (this will always use the traditional covariance matrix)
+
+In our cases, currently we can, still use one option to control them, if we are not going to implement residual/contain covariance matrix.
+Of course, we will not have the option to combine the between within degree of freedom + traditional covariance matrix.
+
+```{r}
+library(Matrix)
+
+fit <- mmrm(FEV1 ~ ARMCD + ar1(AVISIT|USUBJID), data = fev_data)
+
+visits <- fit$tmb_data$visits_zero_inds
+sel_mat <- function(vis, n_vis) {
+  ret <- matrix(0, nrow = length(vis), ncol = n_vis)
+  for (i in seq_len(length(vis))) {
+    ret[i, vis[i]] <- 1
+  }
+  return(ret)
+}
+
+sel_matrix <- sel_mat(visits + 1L, 4)
+sel_matrix_i <- lapply(seq_len(length(fit$tmb_data$subject_zero_inds)), function(x) {
+  sel_matrix[(fit$tmb_data$subject_zero_inds[x]+1):(fit$tmb_data$subject_zero_inds[x]+fit$tmb_data$subject_n_visits[x]),, drop = FALSE]
+})
+v <- lapply(seq_len(length(sel_matrix_i)), function(i) {
+  sel_matrix_i[[i]] %*% fit$cov %*% t(sel_matrix_i[[i]])
+})
+vinv <- lapply(v, solve)
+
+vall <- bdiag(vinv)
+x <- fit$tmb_data$x_matrix
+eps <- fit$tmb_data$y_vector - fit$tmb_data$x_matrix %*% fit$beta_est
+bread <- fit$beta_vcov #solve(t(x) %*% vall %*% x)
+
+meat <- lapply(seq_len(length(sel_matrix_i)), function(i) {
+  ii <- (fit$tmb_data$subject_zero_inds[i]+1):(fit$tmb_data$subject_zero_inds[i]+fit$tmb_data$subject_n_visits[i])
+  t(x[ii, , drop = FALSE]) %*% vinv[[i]] %*% eps[ii, , drop = FALSE] %*% t(eps[ii, drop = FALSE]) %*% vinv[[i]] %*% x[ii, , drop = FALSE]
+})
+meat_sum <- Reduce(`+`, meat)
+
+empirical <- bread %*% meat_sum %*% bread
+```
+
+To achieve this, the following is proposed
+
+```{r}
+fit <- mmrm(FEV1 ~ ARMCD + ar1(AVISIT|USUBJID), data = fev_data, method = "empirical")
+```
+
+with this, the covariance matrix are computed.
+
+# Implementations in `nlme` and `clubSandwich`
+
+To run a similar model, we can use `gls` to create a AR1 mixed model.
+Then we can use `vcovCR` to obtain the empirical covariance matrix.
+Please note that `type = "CR0"` is used to match SAS results(SAS do not do small sample size correction).
```

Comment:
what about the `"CR3"` type? I guess some people could want that ("leave-one-cluster-out jackknife variance estimator (Bell & McCaffrey, 2002)")

Reply:
it this is required, I think we should split the arguments now for finer control. e.g., cov = "empirical"/"empirical_jackknife"/"traditional"/"kenward roger", ddfm = "between within"/"satterthwaite"/"kenward-roger" , and if ddfm is sat or kr, the cov will be changed to "traditional" and "kr adjusted" .

Reply:
or we can add method "empirical jackknife" as one of the supported method

Reply:
ok split might be good then I think. since folks will ask for other CR types. maybe we can keep `method` and add `cov`?

Reply:
ok; will proceed like that; method is for degree of freedom; cov is for "empirical", "empirical jackknife", "traditional"(I wonder if we should have better names for this traditional, $(X^TV^-1X)^-1$ ), "kenward-roger"

Reply:
maybe "asymptotic" instead of traditional?